BarChart: Add ad-hoc filters to tooltip

samjewell opened the pull request:
**What is this feature?**

Allow AdHoc filters to be set from Tooltips when clicking on bars in the BarChart panel.

This is early, experimental work. We want to get it in front of a few users quickly, while we continue to iterate at the same time. See below for the immediate next steps I plan to make in developing this.

For now the UI looks like this:

**On hover:**
<img width="300" src="https://github.com/user-attachments/assets/2fa972c4-40b7-4ff4-b72b-874ff824dbc2" />


**After click:**
<img width="200" src="https://github.com/user-attachments/assets/e8742d32-ab70-4494-b251-46dae080354b" />


**After click, when a datalink is present:**
<img width="200" src="https://github.com/user-attachments/assets/322feefd-1e3d-48c7-b0b0-dca68f0752f2" />


**Styled the same as how we style the "Add annotation" button:**
<img width="250" src="https://github.com/user-attachments/assets/8548feb6-967d-4555-a56d-0bae70edf9d9" />


**When multiple dimensions present**, for now just filter on the first of those dimensions. This isn't configurable (at least not yet), nor explained up front, but it's fast to discover the behaviour. This screenshot filters to `State = California`:
<img width="500" src="https://github.com/user-attachments/assets/94fde5a2-4337-4804-a911-5002f66edd6a" />

I discussed this work with @leeoniya. He was excited, but wanted to see this:
- Rolled out to other panels:
   - That use the TimeSeries Tooltip: Trend and TimeSeries Panels
   - That don't us it: Pie Chart and XY maybe?
- He wanted to see the same code everywhere (consistency across the codebase), even if we have it disabled in other panels / places until we need it.

I've done neither of these things yet, but want to do them next. But first (or rather, in parallel) I want to get the code-change here in front of some users as soon as I can, to get some early feedback 🙏 

Additionally I have some of my own improvements in mind too, but again will follow after this PR. Some ideas:
- Small things, hopefully coming soon:
   - Some refactorings and some tests around the component rendering
   - Renaming and moving some code around, now that AdHoc filters aren't specific to the Table panel only
   - Consider trying out a "One-Click" behaviour, where the first click applies the filter. Maybe this would be configurable - it's a big change to the UX, where currently a click shows the tooltip and from there you can copy data.
- Bigger things, coming later if at all:
   - Include "Filter out value" alongside "filter for value"
   - Shift-click or cmd-click when ad-hoc filtering, to select multiple or add extra values to the selection



**Why do we need this feature?**

We're trying to implement a "Panel to Panel" filtering behaviour, similar to what is possible in Tableau and Looker. We're using AdHoc filters, as these scale well across panels and across dimensions in the data.

We expect this PR will work well in tandem with #108011 to allow panel-to-panel filtering of data coming from BigQuery. 

**Who is this feature for?**

Everyone, but we're targeting some internal Grafana Labs users of Grafana for BI first.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

You can import this dashboard.json to test this out yourself locally:
[AdHocInTooltipsDashboard.json](https://github.com/user-attachments/files/21395916/AdHocInTooltipsDashboard.json)

To get the Prometheus panels populated with data, please run Prometheus locally, and configure a Prom data source on your Grafana instance pointing at that running Prometheus.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] ~~The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.~~

samjewell committed:
Implement AdHoc filter button on BarChart panel

samjewell commented:
Some other refactorings I could make if this work makes its way to production:

1. Move `export type AdHocFilterItem = { key: string; value: string; operator: AdHocFilterOperator };` from both:
   - `/packages/grafana-ui/src/components/Table/TableNG/types.ts#L26`
   - `/packages/grafana-ui/src/components/Table/types.ts#L42`

   Into `@grafana/data`

2.

samjewell committed:
Add feature toggle: `adhocFiltersInTooltips`

samjewell committed:
Gate the feature behind the toggle

samjewell committed:
Fix lint error: can't access feature toggles from inside `grafana-ui`

So we just control the toggle from the TimeSeriesTooltip component only

samjewell committed:
Fix another lint error

samjewell commented on packages/grafana-ui/src/components/VizTooltip/VizTooltipFooter.tsx:
```diff
@@ -88,6 +89,13 @@ export const VizTooltipFooter = ({ dataLinks, actions = [], annotate }: VizToolt
     <div className={styles.wrapper}>
       {!hasOneClickAction && renderDataLinks(dataLinks, styles)}
       {!hasOneClickLink && renderActions(actions, styles)}
+      {!hasOneClickLink && !hasOneClickAction && onFilterClick && (
```
There's some logic here which is untested. I wanted to see how the "Add annotation" button (and the logic that controls its inclusion) was tested, so I removed it to see which tests fail. No tests failed, meaning it has no tests! See this PR's CI run.
So I do plan to investigate adding tests both for this and for the "Add annotation" button

alexjonspencer1 commented:
🏆 Looking good! Left a few comments. 

One thought on the `VizTooltipFooter.tsx` file. I notice we're rendering a variety of buttons that have a lot of the same properties (style classes, etc.). One idea could be to generate an array of buttons that we then iterate in the render (just move all the logic into once place). This would allow us to keep all the render logic in one place in a nice clean way, while also moving the style wrapper to be used universally. Idk if this is making any sense, but here's the code. Just a thought!!

<details>

<summary>Refactor idea ✨ </summary>

### `VizTooltipFooter.tsx`

```Typescript
import { css } from '@emotion/css';
import { useMemo } from 'react';

import { ActionModel, Field, GrafanaTheme2, LinkModel, ThemeSpacingTokens } from '@grafana/data';
import { Trans } from '@grafana/i18n';

import { useStyles2 } from '../../themes/ThemeContext';
import { ActionButton } from '../Actions/ActionButton';
import { Button } from '../Button/Button';
import { DataLinkButton } from '../DataLinks/DataLinkButton';
import { Icon } from '../Icon/Icon';
import { Stack } from '../Layout/Stack/Stack';
import { ResponsiveProp } from '../Layout/utils/responsiveness';

interface VizTooltipFooterProps {
  dataLinks: Array<LinkModel<Field>>;
  actions?: Array<ActionModel<Field>>;
  annotate?: () => void;
  onFilterClick?: () => void;
}

export const ADD_ANNOTATION_ID = 'add-annotation-button';

type RenderOneClickTrans = (title: string) => React.ReactNode;
type RenderItem<T extends LinkModel | ActionModel> = (
  item: T,
  idx: number,
  styles: ReturnType<typeof getStyles>
) => React.ReactNode;

function makeRenderLinksOrActions<T extends LinkModel | ActionModel>(
  renderOneClickTrans: RenderOneClickTrans,
  renderItem: RenderItem<T>,
  itemGap?: ResponsiveProp<ThemeSpacingTokens>
) {
  const renderLinksOrActions = (items: T[], styles: ReturnType<typeof getStyles>) => {
    if (items.length === 0) {
      return;
    }

    const oneClickItem = items.find((item) => item.oneClick === true);

    if (oneClickItem != null) {
      return (
        <Stack direction="column" justifyContent="flex-start" gap={0.5}>
          <span className={styles.oneClickWrapper}>
            <Icon name="info-circle" size="lg" className={styles.infoIcon} />
            {renderOneClickTrans(oneClickItem.title)}
          </span>
        </Stack>
      );
    }

    return (
      <Stack direction="column" justifyContent="flex-start" gap={itemGap}>
        {items.map((item, i) => renderItem(item, i, styles))}
      </Stack>
    );
  };

  return renderLinksOrActions;
}

const renderDataLinks = makeRenderLinksOrActions<LinkModel>(
  (title) => (
    <Trans i18nKey="grafana-ui.viz-tooltip.footer-click-to-navigate">Click to open {{ linkTitle: title }}</Trans>
  ),
  (item, i, styles) => (
    <DataLinkButton link={item} key={i} buttonProps={{ className: styles.dataLinkButton, fill: 'text' }} />
  ),
  0.5
);

const renderActions = makeRenderLinksOrActions<ActionModel>(
  (title) => <Trans i18nKey="grafana-ui.viz-tooltip.footer-click-to-action">Click to {{ actionTitle: title }}</Trans>,
  (item, i) => <ActionButton key={i} action={item} variant="secondary" />
);

export const VizTooltipFooter = ({ dataLinks, actions = [], annotate, onFilterClick }: VizTooltipFooterProps) => {
  const styles = useStyles2(getStyles);
  const hasOneClickLink = useMemo(() => dataLinks.some((link) => link.oneClick === true), [dataLinks]);
  const hasOneClickAction = useMemo(() => actions.some((action) => action.oneClick === true), [actions]);

  const VIZ_TOOLTIP_FOOTER_BUTTONS = [
    {
      shouldRender: !hasOneClickAction,
      render: renderDataLinks(dataLinks, styles),
    },
    {
      shouldRender: !hasOneClickLink,
      render: renderActions(actions, styles),
    },
    {
      shouldRender: !hasOneClickLink && !hasOneClickAction && onFilterClick,
      render: (
        <Button icon="filter" variant="secondary" size="sm" onClick={onFilterClick}>
          <Trans i18nKey="grafana-ui.viz-tooltip.footer-filter-for-value">Filter for value</Trans>
        </Button>
      ),
    },
    {
      shouldRender: !hasOneClickLink && !hasOneClickAction && annotate != null,
      render: (
        <Button icon="comment-alt" variant="secondary" size="sm" id={ADD_ANNOTATION_ID} onClick={annotate}>
          <Trans i18nKey="grafana-ui.viz-tooltip.footer-add-annotation">Add annotation</Trans>
        </Button>
      ),
    },
  ];

  return (
    <div className={styles.wrapper}>
      {VIZ_TOOLTIP_FOOTER_BUTTONS.map(({ shouldRender, render }, index) =>
        shouldRender && render ? (
          <div key={index} className={styles.filterButton}>
            {render}
          </div>
        ) : null
      )}
    </div>
  );
};

const getStyles = (theme: GrafanaTheme2) => ({
  wrapper: css({
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
    padding: theme.spacing(0),
  }),
  filterButton: css({
    borderTop: `1px solid ${theme.colors.border.medium}`,
    padding: theme.spacing(1),
  }),
  dataLinkButton: css({
    cursor: 'pointer',
    '&:hover': {
      textDecoration: 'underline',
      background: 'none',
    },
    padding: 0,
    height: 'auto',
    '& span': {
      whiteSpace: 'normal',
      textAlign: 'left',
    },
  }),
  oneClickWrapper: css({
    display: 'flex',
    alignItems: 'center',
  }),
  infoIcon: css({
    color: theme.colors.primary.main,
    paddingRight: theme.spacing(0.5),
  }),
});

```

</details>

samjewell committed:
Refactor, introducing `.map` to render all buttons that are present

- Simplify the CSS, unifying 3 classes into a single one
- More maintainable - easier to add more buttons later
- Uniform structure - all footer buttons now have consistent DOM structure.

samjewell commented on public/app/plugins/panel/timeseries/TimeSeriesTooltip.tsx:
```diff
@@ -86,11 +91,29 @@ export const TimeSeriesTooltip = ({
+    // create the filter click handler based on hovered series index and row index:
+    const onFilterClick = showFilters
+      ? () => {
+          onAddAdHocFilter?.({
```
I get `'onAddAdHocFilter' is possibly 'undefined'.`. I think we'd need a wider Typescript refactor here.
But thanks for the suggestion

samjewell committed:
Add comment explaining the use of `xField.config.filterable`

alexjonspencer1 reviewed (approved):
LGTM 🏆

leeoniya commented on packages/grafana-ui/src/components/VizTooltip/VizTooltipFooter.tsx:
```diff
@@ -79,21 +76,46 @@ const renderActions = makeRenderLinksOrActions<ActionModel>(
   const hasOneClickLink = useMemo(() => dataLinks.some((link) => link.oneClick === true), [dataLinks]);
   const hasOneClickAction = useMemo(() => actions.some((action) => action.oneClick === true), [actions]);
 
+  const VIZ_TOOLTIP_FOOTER_BUTTONS = [
```
this approach removes optimizations that existed in the previous code, let's tweak this.

jesdavpet replied:
@leeoniya are you recommending to pull just the `renderDataLinks()` and `renderActions()` invocations out of the `.map()` in this PR, or to get rid of the `.map()` altogether?

leeoniya replied:
the way it is now is that it creates an array with objects, each has JSX nodes that can potentially be discarded based on shouldRender.

the previous code avoided allocating anything unless it was necessary to render it, and did not require a wrapping array and wrapper objects.

so yeah, i think we should use the previous approach and just append an adhoc filter rendering condition

leeoniya replied:
since this tooltip re-renders on potentially every mousemove event, this is codepath is hot and should not be wasteful.

leeoniya replied:
for barchart this isnt a big deal, bars are wide and few, and the tooltip core wont rerender the contents when the hovered value is same from one pixel to next.

but this is a shared component, and in timeseries panel it's significant

samjewell committed:
Revert "Refactor, introducing `.map` to render all buttons that are present"

This reverts commit 7c03a18dfe4522830e1061354b5ac22bd722fc3d.

samjewell committed:
Refactor VizTooltipFooter component to unify button styles

- Changed multiple class names to a single `footerButton` class for consistency

leeoniya committed:
Merge branch 'main' into sj/adhoc-filter-barchart

leeoniya committed:
support multiple filters, move xField implementation closer to BarChart

leeoniya committed:
fix

leeoniya committed:
better button?

leeoniya committed:
internal

leeoniya committed:
lint

leeoniya committed:
fix

leeoniya committed:
Merge branch 'main' into sj/adhoc-filter-barchart

samjewell commented on packages/grafana-ui/src/components/VizTooltip/VizTooltipFooter.tsx:
```diff
@@ -114,7 +121,7 @@ const getStyles = (theme: GrafanaTheme2) => ({
     padding: theme.spacing(0),
   }),
-  footerButton: css({
+  footerSection: css({
```
I'm tempted to simplify this further, to `footer: …`

samjewell commented on public/app/plugins/panel/barchart/BarChartPanel.tsx:
```diff
@@ -161,7 +163,40 @@ export const BarChartPanel = (props: PanelProps<Options>) => {
                 vizSeries[0].fields[seriesIdx].getLinks?.({ valueRowIndex: dataIdx }) ?? []
               }
-              render={(u, dataIdxs, seriesIdx, isPinned, dismiss, timeRange2, viaSync, dataLinks) => {
+              getAdHocFilters={(seriesIdx, dataIdx) => {
```
I get a suggestion in my IDE `'seriesIdx' is declared but its value is never read`

Should we remove this?

samjewell commented on public/app/plugins/panel/barchart/BarChartPanel.tsx:
```diff
@@ -161,7 +163,40 @@ export const BarChartPanel = (props: PanelProps<Options>) => {
               }
-              render={(u, dataIdxs, seriesIdx, isPinned, dismiss, timeRange2, viaSync, dataLinks) => {
+              getAdHocFilters={(seriesIdx, dataIdx) => {
+                const xField = vizSeries[0].fields[0];
```
I notice here we're always choosing `visSeries[0]` when we read the first field. I find myself wondering whether we can ever have `vizSeries[1].fields[0]` hold different or inconsistent data here.

When the block of code below was still in `TimeSeriesTooltip.tsx` it relied on `const xField = series.fields[0];`, so I was somehow less worried.

samjewell replied:
Claude confirmed to me that there's no risk here:

> vizSeries is always a single-element array: The prepSeries function in utils.ts always returns const series = [frame] (line 134), regardless of how many input frames it receives.

alexjonspencer1 commented on packages/grafana-ui/src/components/VizTooltip/VizTooltipFooter.tsx:
```diff
@@ -11,12 +11,17 @@ import { DataLinkButton } from '../DataLinks/DataLinkButton';
 interface VizTooltipFooterProps {
   dataLinks: Array<LinkModel<Field>>;
   actions?: Array<ActionModel<Field>>;
+  adHocFilters?: AdHocFilterModel[];
```
Oh yeah, probably! Usually my linter complains about this 🤔

samjewell replied:
Arg! 

> Array type using 'Array<AdHocFilterModel>' is forbidden for simple types. Use 'AdHocFilterModel[]' instead

https://github.com/grafana/grafana/actions/runs/16602959618/job/46967370743?pr=107874

Thanks though

samjewell commented on packages/grafana-ui/src/components/VizTooltip/VizTooltipFooter.tsx:
```diff
@@ -88,8 +94,19 @@ export const VizTooltipFooter = ({ dataLinks, actions = [], annotate }: VizToolt
+      {!hasOneClickLink && !hasOneClickAction && adHocFilters.length > 0 && (
+        <div className={styles.footerSection}>
+          {adHocFilters.map((item) => (
+            <Button icon="filter" variant="secondary" size="sm" onClick={item.onClick}>
```
Claude is telling me we should add React Keys here:

```suggestion
            <Button key={index} icon="filter" variant="secondary" size="sm" onClick={item.onClick}>
```

Or it thinks this could be even better:
```suggestion
            <Button key={`${item.key}-${item.value}`} icon="filter" variant="secondary" size="sm" onClick={item.onClick}>
```

samjewell committed:
Add key to mapped buttons for improved rendering

And remove a code comment

samjewell committed:
Indicate unused function-param with underscore
